[unit test] Timed out after 10.001s.

### Which jobs are flaking?

- periodic-cluster-api-test-mink8s-main
- periodic-cluster-api-test-release-1-10

Failure:

```
Failed
=== RUN   TestExtensionReconciler_Reconcile
I0601 22:44:22.407311   32510 feature_gate.go:385] feature gates: {map[ClusterTopology:true]}
I0601 22:44:22.407441   32510 feature_gate.go:385] feature gates: {map[ClusterTopology:true RuntimeSDK:true]}
I0601 22:44:22.408648   32510 ???:1] "http: TLS handshake error from 127.0.0.1:58002: EOF"
I0601 22:44:22.532986   32510 reflector.go:430] "Caches populated" logger="controller-runtime.cache" type="*v1beta2.ExtensionConfig" reflector="pkg/mod/k8s.io/client-go@v0.33.1/tools/cache/reflector.go:285"
    extensionconfig_controller_test.go:122: 
        Timed out after 10.001s.
        The function passed to Eventually failed at /home/prow/go/src/sigs.k8s.io/cluster-api/exp/runtime/internal/controllers/extensionconfig_controller_test.go:121 with:
        Expected
            <int64>: 1
        to equal
            <int64>: 2
--- FAIL: TestExtensionReconciler_Reconcile (10.81s)
```

triage link: https://storage.googleapis.com/k8s-triage/index.html?text=The%20function%20passed%20to%20Eventually%20failed%20at%20%2Fhome%2Fprow%2Fgo%2Fsrc%2Fsigs.k8s.io%2Fcluster-api%2Fexp%2Fruntime%2Finternal%2Fcontrollers%2Fextensionconfig_controller_test.go&job=.*-cluster-api-.*&xjob=.*-provider-.*%7C.*-operator-.* 

Failure link: https://prow.k8s.io/view/gs/kubernetes-ci-logs/logs/periodic-cluster-api-test-mink8s-main/1929305699737669632 

### Which tests are flaking?

sigs.k8s.io/cluster-api/exp/runtime/internal/controllers TestExtensionReconciler_Reconcile

### Since when has it been flaking?

first failure on triage board [19/05/2025](https://prow.k8s.io/view/gs/kubernetes-ci-logs/logs/periodic-cluster-api-test-release-1-10/1924297096265994240)

Not sure when it started.

### Testgrid link

https://testgrid.k8s.io/sig-cluster-lifecycle-cluster-api#capi-test-mink8s-main

### Reason for failure (if possible)

_No response_

### Anything else we need to know?

_No response_

### Label(s) to be applied

/kind flake
One or more /area label. See https://github.com/kubernetes-sigs/cluster-api/labels?q=area for the list of labels.

--- a/exp/runtime/internal/controllers/extensionconfig_controller_test.go
+++ b/exp/runtime/internal/controllers/extensionconfig_controller_test.go
@@ -69,7 +69,7 @@ func TestExtensionReconciler_Reconcile(t *testing.T) {
 	g.Expect(runtimehooksv1.AddToCatalog(cat)).To(Succeed())
 
 	r := &Reconciler{
-		Client:        env.GetClient(),
+		Client:        env.GetAPIReader().(client.Client),
 		APIReader:     env.GetAPIReader(),
 		RuntimeClient: runtimeClient,
 	}
@@ -102,7 +102,7 @@ func TestExtensionReconciler_Reconcile(t *testing.T) {
 	t.Run("successful reconcile and discovery on ExtensionConfig create", func(*testing.T) {
 		// Warm up the registry before trying reconciliation again.
 		warmup := &warmupRunnable{
-			Client:        env.GetClient(),
+			Client:        env.GetAPIReader().(client.Client),
 			APIReader:     env.GetAPIReader(),
 			RuntimeClient: runtimeClient,
 		}

--- a/exp/runtime/internal/controllers/extensionconfig_controller_test.go
+++ b/exp/runtime/internal/controllers/extensionconfig_controller_test.go
@@ -119,18 +119,10 @@ func TestExtensionReconciler_Reconcile(t *testing.T) {
 			pausedCondition := conditions.Get(conf, clusterv1.PausedCondition)
 			g.Expect(pausedCondition).ToNot(BeNil())
 			g.Expect(pausedCondition.ObservedGeneration).To(Equal(conf.Generation))
-		}).WithTimeout(15 * time.Second).WithPolling(100 * time.Millisecond).Should(Succeed())
+		}).WithTimeout(10 * time.Second).WithPolling(100 * time.Millisecond).Should(Succeed())
 
 		_, err = r.Reconcile(ctx, ctrl.Request{NamespacedName: util.ObjectKey(extensionConfig)})
 		g.Expect(err).ToNot(HaveOccurred())
-
-		// Wait for discovery to complete before checking final conditions
-		g.Eventually(func(g Gomega) {
-			updatedConfig := &runtimev1.ExtensionConfig{}
-			g.Expect(env.GetAPIReader().Get(ctx, util.ObjectKey(extensionConfig), updatedConfig)).To(Succeed())
-			v1beta2Conditions := updatedConfig.GetConditions()
-			g.Expect(len(v1beta2Conditions)).To(BeNumerically(">=", 2))
-		}).WithTimeout(10 * time.Second).WithPolling(100 * time.Millisecond).Should(Succeed())
 
 		config := &runtimev1.ExtensionConfig{}
 		g.Expect(env.GetAPIReader().Get(ctx, util.ObjectKey(extensionConfig), config)).To(Succeed())

--- a/exp/runtime/internal/controllers/extensionconfig_controller_test.go
+++ b/exp/runtime/internal/controllers/extensionconfig_controller_test.go
@@ -119,10 +119,18 @@ func TestExtensionReconciler_Reconcile(t *testing.T) {
 			pausedCondition := conditions.Get(conf, clusterv1.PausedCondition)
 			g.Expect(pausedCondition).ToNot(BeNil())
 			g.Expect(pausedCondition.ObservedGeneration).To(Equal(conf.Generation))
-		}).WithTimeout(10 * time.Second).WithPolling(100 * time.Millisecond).Should(Succeed())
+		}).WithTimeout(15 * time.Second).WithPolling(100 * time.Millisecond).Should(Succeed())
 
 		_, err = r.Reconcile(ctx, ctrl.Request{NamespacedName: util.ObjectKey(extensionConfig)})
 		g.Expect(err).ToNot(HaveOccurred())
+
+		// Wait for discovery to complete before checking final conditions
+		g.Eventually(func(g Gomega) {
+			updatedConfig := &runtimev1.ExtensionConfig{}
+			g.Expect(env.GetAPIReader().Get(ctx, util.ObjectKey(extensionConfig), updatedConfig)).To(Succeed())
+			v1beta2Conditions := updatedConfig.GetConditions()
+			g.Expect(len(v1beta2Conditions)).To(BeNumerically(">=", 2))
+		}).WithTimeout(10 * time.Second).WithPolling(100 * time.Millisecond).Should(Succeed())
 
 		config := &runtimev1.ExtensionConfig{}
 		g.Expect(env.GetAPIReader().Get(ctx, util.ObjectKey(extensionConfig), config)).To(Succeed())

